src/infer_type.rs

```diff
@@ -0,0 +1,131 @@
+use crate::types::{Dimension, GeometryType};
+
+const POINT: &str = "POINT";
+const LINESTRING: &str = "LINESTRING";
+const POLYGON: &str = "POLYGON";
+const MULTIPOINT: &str = "MULTIPOINT";
+const MULTILINESTRING: &str = "MULTILINESTRING";
+const MULTIPOLYGON: &str = "MULTIPOLYGON";
+const GEOMETRYCOLLECTION: &str = "GEOMETRYCOLLECTION";
+
+/// Infer the geometry type and dimension from an input WKT string slice.
+///
+/// An `EMPTY` WKT object will return `None` in place of the dimension.
+///
+/// ```
+/// use wkt::infer_type;
+/// use wkt::types::{Dimension, GeometryType};
+///
+/// assert_eq!(
+///     infer_type("POINT (10 20.1)").unwrap(),
+///     (GeometryType::Point, Some(Dimension::XY))
+/// );
+///
+/// assert_eq!(
+///     infer_type("POINT EMPTY").unwrap(),
+///     (GeometryType::Point, None)
+/// );
+/// ```
+pub fn infer_type(input: &str) -> Result<(GeometryType, Option<Dimension>), String> {
+    if let Some((prefix, _suffix)) = input.split_once("(") {
+        let prefix = prefix.to_uppercase();
+
+        let (geom_type, dim_str) = if let Some(dim_str) = prefix.strip_prefix(POINT) {
+            (GeometryType::Point, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(LINESTRING) {
+            (GeometryType::LineString, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(POLYGON) {
+            (GeometryType::Polygon, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(MULTIPOINT) {
+            (GeometryType::MultiPoint, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(MULTILINESTRING) {
+            (GeometryType::MultiLineString, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(MULTIPOLYGON) {
+            (GeometryType::MultiPolygon, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(GEOMETRYCOLLECTION) {
+            (GeometryType::GeometryCollection, dim_str)
+        } else {
+            return Err(format!("Unsupported WKT prefix {}", prefix));
+        };
+
+        let dim = if dim_str.contains("ZM") {
+            Dimension::XYZM
+        } else if dim_str.contains("Z") {
+            Dimension::XYZ
+        } else if dim_str.contains("M") {
+            Dimension::XYM
+        } else {
+            Dimension::XY
+        };
+
+        Ok((geom_type, Some(dim)))
+    } else {
+        let input = input.to_uppercase();
```

Comment:
It'd be nice to avoid allocation by doing a case insensitive match, rather than copying to a String, but usually these are pretty small, so no biggie. Take it or leave it.

Reply:
1. This is in the block that doesn't contain a `(`, so this should only ever be called on a string as long as `MULTIPOLYGON EMPTY` or so. (Unless the input is invalid)
2. I still agree that I'd prefer not having the allocation. But one of the existing tests was testing the validity of `point empty`, and wasn't sure the best way to handle this. I suppose we could match first on the original input, and if none of those match, cast to uppercase and do it again?

Reply:
I think it's not worth complicating the code much. If it's not a couple word change, then I'd leave it as is.

Reply:
Well, actually, this PR currently only supports lower case WKT input in the `EMPTY` case. So we have a test that `point EMPTY` passes, but `point (10, 20)` would currently fail.

I figure both arms should do the same. Should we remove lower case support from the `EMPTY` arm or add it to the primary branch?

Reply:
> Well, actually, this PR currently only supports lower case WKT input in the EMPTY case.

I don't understand what you mean. To me it looks like there is already similar "to_uppercase" logic in the other branch.

Since I seem to be lost, rather than talking about the structure of your particular code, I'll explain the behavior that I think makes sense:

In all cases:
- it should probably work with input that our library can parse ( which includes being case insensitive and being resilient to whitespace sprinkled before/after/between tokens)
- it should avoid heap allocation if it doesn't overly complicate the code

Reply:
Oops you're right, I missed that line 😅 https://github.com/georust/wkt/pull/119/files#diff-14a520b6386f7584931dd3fd2cc7a098cdc7137c83cf69699e115cd74bba13ceR31